File: 432. All O`one Data Structure/Solution.cpp
#include <list>

using namespace std;

class AllOne {
private:
    list<pair<int, unordered_set<string>>> keysOf;
    unordered_map<string, keysOf::iterator> valuesOf;
public:
    /** Initialize your data structure here. */
    AllOne() {

    }

    /** Inserts a new key < Key > with value 1. Or increments an existing key by 1. */
    void inc(string &key) {
        if (list.empty()) {
            keysOf.insert(make_pair<int, unordered_set<string>>(1, unordered_set<string>()));
            (--keysOf.end())->insert(key);
            valuesOf[string] = keysOf.cbegin();
        } else {
            keysOf::iterator itr = valuesOf[key];
            if (++itr == keysOf.end())
                keysOf.insert(make_pair<int, unordered_set<string>>((--itr)->first + 1, unordered_set<string>()))
            (++itr)->second.insert(key);
            (--itr)->second.remove(key);
            if (itr->second.empty()) 
                keysOf.erase(itr);
        }
    }

    /** Decrements an existing key by 1. If Key's value is 1, remove it from the data structure. */
    void dec(string &key) {
        if (!valuesOf.count(key)) return;
        
    }

    /** Returns one of the keys with maximal value. */
    string getMaxKey() {
        if (keysOf.empty()) return "";
        else return *keysOf.rbegin();
    }

    /** Returns one of the keys with Minimal value. */
    string getMinKey() {
        if (keysOf.empty()) return "";
        else return *keysOf.begin();
    }
}

/**
    * Your AllOne object will be instantiated and called as such:
    * AllOne obj = new AllOne()
    * obj.inc(key)
    * obj.dec(key)
    * string param_3 = obj.getMaxKey()
    * string param_4 = obj.getMinKey()
    */

File: 432. All O`one Data Structure/Solution.py
class AllOne(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.kcnt = collections.defaultdict(int)
        self.kset = [set()]

    def inc(self, key):
        """
        Inserts a new key <Key> with value 1. Or increments an existing key by 1.
        :type key: str
        :rtype: void
        """
        self.kcnt[key] += 1
        cnt = self.kcnt[key]
        if cnt >= len(self.kset):
            self.kset.append(set())
        self.kset[cnt].add(key)
        if cnt - 1 > 0:
            self.kset[cnt - 1].remove(key)

    def dec(self, key):
        """
        Decrements an existing key by 1. If Key's value is 1, remove it from the data structure.
        :type key: str
        :rtype: void
        """
        self.kcnt[key] -= 1
        cnt = self.kcnt[key]
        if key in self.kset[cnt + 1]:
            self.kset[cnt + 1].remove(key)
        if cnt <= 0:
            del self.kcnt[key]
        else:
            self.kset[cnt].add(key)
        if len(self.kset) > 1 and len(self.kset[-1]) == 0:
            self.kset.pop()

    def getMaxKey(self):
        """
        Returns one of the keys with maximal value.
        :rtype: str
        """
        if len(self.kset) <= 1:
            return ""
        else:
            for k in self.kset[-1]:
                return k

    def getMinKey(self):
        """
        Returns one of the keys with Minimal value.
        :rtype: str
        """
        if len(self.kset) <= 1:
            return ""
        else:
            for k in self.kset[1]:
                return k


# Your AllOne object will be instantiated and called as such:
# obj = AllOne()
# obj.inc(key)
# obj.dec(key)
# param_3 = obj.getMaxKey()
# param_4 = obj.getMinKey()
